cpp/oneapi/dal/algo/kmeans/backend/gpu/train_kernel_lloyd_dense_dpc.cpp

```diff
@@ -96,65 +95,153 @@ struct train_kernel_gpu<Float, method::lloyd_dense, task::clustering> {
+                                                  arr_objective_function,
+                                                  { assign_event });
+            centroids_event = partial_reduce_centroids<Float>(queue,
+                                                              arr_data,
+                                                              arr_labels,
+                                                              cluster_count,
+                                                              part_count,
+                                                              arr_partial_centroids,
+                                                              { count_event });
+            centroids_event = merge_reduce_centroids<Float>(queue,
+                                                            arr_counters,
+                                                            arr_partial_centroids,
+                                                            part_count,
+                                                            arr_centroids,
+                                                            { count_event, centroids_event });
+            count_empty_clusters(queue,
+                                 cluster_count,
+                                 arr_counters,
+                                 arr_num_empty_clusters,
+                                 { count_event })
+                .wait_and_throw();
+
+            std::int64_t num_candidates = arr_num_empty_clusters.to_host(queue).get_data()[0];
+            sycl::event find_candidates_event;
+            if (num_candidates > 0) {
+                find_candidates_event = find_candidates<Float>(queue,
+                                                               arr_closest_distances,
+                                                               num_candidates,
+                                                               arr_candidate_indices,
+                                                               arr_candidate_distances);
+            }
+
+            objective_function_event.wait_and_throw();
+            sycl::event::wait({ find_candidates_event, objective_function_event });
+            Float objective_function = arr_objective_function.to_host(queue).get_data()[0];
+
+            bk::event_vector candidate_events;
+            if (num_candidates > 0) {
+                candidate_events = fill_empty_clusters(queue,
+                                                       arr_data,
+                                                       arr_counters,
+                                                       arr_candidate_indices,
+                                                       arr_candidate_distances,
+                                                       arr_centroids,
+                                                       arr_labels,
+                                                       objective_function,
```

Comment:
If objective function is updated by this function, I suggest to return its values explicitly using `std::tuple`

Reply:
Done